File: demo/KSYLiveDemo/KSYDemoUI/KSYAudioCtrlView.m
//
//  KSYAudioCtrlView.m
//  KSYGPUStreamerDemo
//
//  Created by 孙健 on 16/6/24.
//  Copyright © 2016年 ksyun. All rights reserved.
//

#import <GPUImage/GPUImage.h>
#import "KSYPresetCfgView.h"
#import "KSYAudioCtrlView.h"
#import "KSYNameSlider.h"
@interface KSYAudioCtrlView() {
    
}

@property UILabel       * lblPlayCapture;
@property UILabel       * lblMuteSt;
@property UILabel       * lblReverb;
@property UILabel       * lblVPIO;
@property UILabel       * lblStereo;
@end
@implementation KSYAudioCtrlView

- (id)init{
    self = [super init];
    // 混音音量
    _micVol = [self addSliderName:@"麦克风音量" From:0.0 To:2.0 Init:0.9];
    _bgmVol = [self addSliderName:@"背景乐音量"  From:0.0 To:2.0 Init:0.5];
    _bgmMix = [self addSwitch:YES];


    _micInput = [self addSegCtrlWithItems:@[ @"内置mic", @"耳麦", @"蓝牙mic"]];
    [self initMicInput];
    
    _lblAudioOnly    = [self addLable:@"纯音频推流"]; // 关闭视频
    _swAudioOnly     = [self addSwitch:NO]; // 关闭视频
    _lblMuteSt       = [self addLable:@"静音推流"];
    _muteStream      = [self addSwitch:NO];
    
    _lblStereo     = [self addLable:@"立体声推流"];
    _stereoStream  = [self addSwitch:NO];
    _lblReverb  = [self addLable:@"混响"];
    _reverbType = [self addSegCtrlWithItems:@[@"关闭", @"录影棚",
                                              @"演唱会",@"KTV",@"小舞台"]];
    _lblPlayCapture = [self addLable:@"耳返"];
    _swPlayCapture  = [self addSwitch:NO];
    _playCapVol= [self addSliderName:@"耳返音量"  From:0.0 To:1.0 Init:0.5];
    _effectType  = [self addSegCtrlWithItems:@[@"关闭变声",@"大叔", @"萝莉", @"庄严", @"机器人"]];
    return self;
}
- (void)layoutUI{
    [super layoutUI];
    self.btnH = 30;

    [self putRow1:_micVol];
    [self putSlider:_bgmVol
          andSwitch:_bgmMix];
    [self putRow1:_micInput];
    [self putLable:_lblReverb andView:_reverbType];
    id nu = [NSNull null];
    [self putRowFit:@[_lblAudioOnly,_swAudioOnly,
                      nu, _lblMuteSt,_muteStream,
                      nu,_lblPlayCapture,_swPlayCapture] ];
    [self putRow1:_playCapVol];
    [self putRow1:_effectType];
    [self putRowFit:@[_lblStereo, _stereoStream]];
}
- (void) initMicInput {
    BOOL bHS = [AVAudioSession isHeadsetInputAvaible];
    BOOL bBT = [AVAudioSession isBluetoothInputAvaible];
    [_micInput setEnabled:YES forSegmentAtIndex:1];
    [_micInput setEnabled:YES forSegmentAtIndex:2];
    if (!bHS){
        [_micInput setEnabled:NO forSegmentAtIndex:1];
    }
    if (!bBT){
        [_micInput setEnabled:NO forSegmentAtIndex:2];
    }
}

static int micType2Int( KSYMicType t) {
    if (t == KSYMicType_builtinMic){
        return 0;
    }
    else if (t == KSYMicType_headsetMic){
        return 1;
    }
    else if (t == KSYMicType_bluetoothMic){
        return 2;
    }
    return 0;
}

static KSYMicType int2MicType( int t) {
    if (t == 0){
        return KSYMicType_builtinMic;
    }
    else if (t == 1){
        return KSYMicType_headsetMic;
    }
    else if (t == 2){
        return KSYMicType_bluetoothMic;
    }
    return KSYMicType_builtinMic;
}

@synthesize  micType = _micType;
- (void) setMicType:(KSYMicType)micType{
    _micType = micType;
    _micInput.selectedSegmentIndex = micType2Int(micType);
}

- (KSYMicType) micType{
    _micType = int2MicType((int)_micInput.selectedSegmentIndex);
    return _micType;
}
@synthesize audioEffect = _audioEffect;
- (void) setAudioEffect:(KSYAudioEffectType)audioEffect {
    _audioEffect = audioEffect;
    if (_audioEffect < 5 ) {
        _effectType.selectedSegmentIndex  = (NSInteger) _audioEffect;
    }
}
- (KSYAudioEffectType) audioEffect {
    _audioEffect =  _effectType.selectedSegmentIndex;
    return _audioEffect;
}
@end

File: demo/KSYLiveDemo_Swift/KSYDemoUI_Swift/subviews/KSYAudioCtrlView.swift
//
//  KSYAudioCtrlView.swift
//  KSYLiveDemo_Swift
//
//  Created by iVermisseDich on 17/1/10.
//  Copyright © 2017年 com.ksyun. All rights reserved.
//

import UIKit

/**
 音频控制相关
 
 主要增加的功能如下:
 1. 混音
 2. 输入音频设备选择
 3. 混响类型选择
 4. 耳返
 */

class KSYAudioCtrlView: KSYUIView {

    var micVol: KSYNameSlider?          /// 混音时, 麦克风的比例
    var bgmVol: KSYNameSlider?          /// 混音时, 背景音乐的比例
    var bgmMix: UISwitch?               /// 混音时, 背景音乐是否混入
    
    var swAudioOnly: UISwitch?          /// 纯音频推流开关 ( 纯音频 == 关闭视频 )
    var muteStream: UISwitch?           /// 静音推流开关 ( 发送音量为0的音频数据 )
    
    var micInput: UISegmentedControl?   /// 音频输入设备选择(话筒, 有限耳麦 或 蓝牙耳麦)
    
    var reverbType: UISegmentedControl? /// 混响类型选择
    var swPlayCapture: UISwitch?        /// 耳返 (本地直接播放采集到的声音) (请戴耳机之后再使用本功能)
    var playCapVol: KSYNameSlider?      /// 本地播放的音量
    var _micType: KSYMicType?
    var micType: KSYMicType? {
        get{
            return getMicType()
        }
        set{
            _micType = newValue
            self.micInput?.selectedSegmentIndex = Int(newValue?.rawValue ?? 0)
        }
    }
    
    func getMicType() -> KSYMicType {
        return KSYMicType.init(rawValue: UInt(micInput!.selectedSegmentIndex))!
    }
    
    private
    var lblPlayCapture: UILabel?
    var lblAudioOnly: UILabel?
    var lblMuteSt: UILabel?
    var lblReverb: UILabel?
    
    override init(withParent pView: KSYUIView) {
        super.init(withParent: pView)
        micVol = addSlider(name: "麦克风音量", from: 0.0, to: 2.0, initV: 0.9)
        bgmVol = addSlider(name: "背景乐音量", from: 0.0, to: 2.0, initV: 0.5)
        bgmMix = addSwitch(on: true)
        
        micInput = addSegCtrlWithItems(items: ["内置mic", "耳麦", "蓝牙mic"])
        initMicInput()
        
        lblAudioOnly = addLabel(title: "纯音频推流")  // 关闭视频
        swAudioOnly = addSwitch(on: false)          // 关闭视频
        lblMuteSt = addLabel(title: "静音推流")
        muteStream = addSwitch(on: false)
        
        lblReverb = addLabel(title: "混响")
        reverbType = addSegCtrlWithItems(items: ["关闭", "录影棚",
                                                  "演唱会","KTV","小舞台"])
        
        lblPlayCapture = addLabel(title: "耳返")
        swPlayCapture = addSwitch(on: false)
        playCapVol = addSlider(name: "耳返音量", from: 0.0, to: 1.0, initV: 0.5)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutUI() {
        super.layoutUI()
        btnH = 30
        
        putRow1(subV: micVol!)
        putSlider(sl: bgmVol!, andSwitch: bgmMix!)
        putRow1(subV: micInput!)
        putRow(subV: [lblAudioOnly!, swAudioOnly!, lblMuteSt!, muteStream!])
        putLabel(lbl: lblReverb!, andView: reverbType!)
        
        // tip: Array cannot contains NSNull
        let arr: NSArray = NSArray.init(objects: NSNull(), NSNull(), lblPlayCapture!, swPlayCapture!)
        putRow(subV2: arr)
        putRow1(subV: playCapVol)
    }
    
    /// 初始化mic选择控件
    func initMicInput() {
        let bHS = KSYAVAudioSession.isHeadsetInputAvaible()
        let bBT = KSYAVAudioSession.isBluetoothInputAvaible()
        
        micInput?.setEnabled(true, forSegmentAt: 1)
        micInput?.setEnabled(true, forSegmentAt: 2)
        
        if !bHS {
            micInput?.setEnabled(false, forSegmentAt: 1)
        }
        if !bBT {
            micInput?.setEnabled(false, forSegmentAt: 2)
        }
    }
    
}
